cmd/log/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+///
+/// Extract the timestamp from the line
+///
+fn extract_timestamp(l: &str) -> Result<u64, ParseIntError> {
+    let mut line = l.split(':');
+    // the first element before the ':' should be the timestamp
+    let hopefully_number_part = line.next().unwrap();
+    hopefully_number_part.parse()
+}
+
+fn stringbuf_monitor(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    delay: Option<u64>,
+    once: bool,
+) -> Result<()> {
+    let mut buf: Vec<u8> = vec![];
+    buf.resize_with(ringbuf_var.size, Default::default);
+
+    let mut update_stringbuf = || -> Stringbuf {
+        let _info = core.op_start().unwrap();
+        core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+        core.op_done().unwrap();
+
+        // There are two possible shapes of ringbufs, depending on the age of the
+        // firmware.
+        // - Raw Ringbuf that is not wrapped by anything.
+        // - Safe Ringbuf that is inside a StaticCell.
+        //
+        // Here we will attempt to handle them both -- first raw, then fallback.
+        let ringbuf_val: Value = Value::Struct(
+            reflect::load_struct(hubris, &buf, definition, 0).unwrap(),
+        );
+
+        let cell: StaticCell = StaticCell::from_value(&ringbuf_val).unwrap();
+        Stringbuf::from_value(&cell.cell.value).unwrap()
+    };
+
+    // use a BTree to keep a sorted hashmap of all the log entries,  this prevents reprinting log
+    // lines as well as keeping them ordered.
+    let mut full_log = BTreeMap::new();
+
+    //
+    // TODO
+    // set ctrl c handler before looping
+    // this will ensure that op_done is always called even if interrupted
+    // see jira https://rivosinc.atlassian.net/browse/SW-440
+    //
+
+    loop {
+        //
+        // Read the log from the core and convert to a big string
+        //
+        let stringbuf = update_stringbuf();
+        let buffer = stringbuf.buffer.as_slice();
+        let log = str::from_utf8(buffer);
+        if let Err(_err) = log {
+            println!("Invalid Log: {:?}", stringbuf);
+        }
+
+        //
+        // split log file up into lines based on null delimeter,
+        // also count the characters
+        //
+        let log = log.unwrap();
+        let mut lines: Vec<&str> = log.split('\0').collect();
+        let char_counts: Vec<usize> =
+            lines.iter().map(|line| line.len()).collect();
+
+        //
+        // remove the line that is currently being overwritten
+        //
+        let last_idx: usize = stringbuf.last.unwrap() as usize;
+        let last_char = buffer[last_idx];
+        if last_char != 0 && last_char != b'\r' {
+            let mut total_count = 0;
+            for (char_count, (i, _line)) in
+                zip(char_counts, lines.clone().iter().enumerate())
+            {
```

Comment:
Would `lines.truncate(lines.len() - 1)` not work here?

Reply:
`lines` is not necessarily in order, so we have to find the last entry using `last` from the `stringbuf` structure.